### PR TITLE
Support building image for ARM

### PR DESCRIPTION
**Pull request by buffalu**
- Add multi-platform build for CI.
- Fix dockerfile to pin platform used in docker images
- Add building on big ubuntu machine.
- Add rust-toolchain.toml that pins cargo version used to build (no fighting between local + CI cargo images).
- Clippy fix

### CODE REVIEW

**Review by 0xEdgar — APPROVED** [review]

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,72 @@
+name: release
+
+on:
+  push:
+    #    branches:
```
Comment (0xEdgar) [inline .github/workflows/release.yml:5]: comment?

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -1,10 +1,11 @@
 name: test
 
-on: push
+on:
+  pull_request:
 
 jobs:
   test:
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-20.04-16c-64g
```
Comment (0xEdgar) [inline .github/workflows/test.yml:8]: latest breaks tests?

File: Cargo.toml [hunk 3]
```diff
@@ -22,12 +22,12 @@ rand_07 = { package = "rand", version = "~0.7" }
...
+solana-metrics = "~1.14"
+solana-net-utils = "~1.14"
+solana-perf = "~1.14"
+solana-sdk = "~1.14"
+solana-streamer = "~1.14"
 thiserror = "1.0.40"
 tokio = "1"
-tonic = { version = "0.9.2", features = ["tls", "tls-roots", "tls-webpki-roots"] }
\ No newline at end of file
+tonic = { version = "0.9.2", features = ["tls", "tls-roots", "tls-webpki-roots"] }
```
Comment (0xEdgar) [inline Cargo.toml:33]: nit \n

File: Dockerfile [hunk 4]
```diff
@@ -1,5 +1,5 @@
 # syntax=docker/dockerfile:1.4.0
-FROM rust:1.66-slim-bullseye as builder
+FROM --platform=linux/amd64 rust:1.66-slim-bullseye as builder
```
Comment (0xEdgar) [inline Dockerfile:2]: does this build all future images as amd64, including on prod? or is this conditionally built somehow

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        with:
+          username: ${{ secrets.DOCKERHUB_USER }}
+          password: ${{ secrets.DOCKERHUB_PWD }}
+
+      - name: Build and push
+        uses: docker/build-push-action@v4
+        with:
+          context: .
+          push: true
+          tags: jitolabs/jito-shredstream-proxy:${{github.sha}}
```
Comment (esemeniuc) [inline .github/workflows/release.yml:53]: no version tag?

File: Dockerfile [hunk 6]
```diff
@@ -8,20 +8,16 @@ ENV HOME=/home/root
 WORKDIR $HOME/app
 COPY . .
 
-RUN --mount=type=cache,mode=0777,target=/home/root/app/target \
-    --mount=type=cache,mode=0777,target=/usr/local/cargo/registry \
-    --mount=type=cache,mode=0777,target=/usr/local/cargo/git \
-    RUSTFLAGS="-C target-cpu=x86-64-v3" cargo build --release && cp target/release/jito-* ./
+RUN cargo build --release
```
Comment (esemeniuc) [inline Dockerfile:11]: any chance to save caching?

File: src/main.rs [hunk 7]
```diff
@@ -141,10 +141,12 @@ pub enum ShredstreamProxyError {
 }
 
 fn resolve_hostname_port(hostname_port: &str) -> io::Result<(SocketAddr, String)> {
-    let socketaddr = hostname_port.to_socket_addrs()?.next().ok_or(Error::new(
-        ErrorKind::AddrNotAvailable,
-        format!("Could not find destination {hostname_port}"),
-    ))?;
+    let socketaddr = hostname_port.to_socket_addrs()?.next().ok_or_else(|| {
```
Comment (esemeniuc) [inline src/main.rs:144]: curious what is the benefit of using `ok_or_else`?

File: .github/workflows/release.yml (hunk 5 above)
Comment (buffalu) [reply .github/workflows/release.yml:53]: oops thats why its not working lol

Comment (buffalu) [reply .github/workflows/release.yml]: bug, fixed

File: Dockerfile (hunk 6 above)
Comment (buffalu) [reply Dockerfile:11]: yeah lemme add this back in

File: src/main.rs (hunk 7 above)
Comment (buffalu) [reply src/main.rs:144]: https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or

> Arguments passed to ok_or are eagerly evaluated; if you are passing the result of a function call, it is recommended to use [ok_or_else](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or_else), which is lazily evaluated.

**Pull request merged**